### ISSUE TITLE
migrate clj-time to JSR-310

### ISSUE THREAD

**Issue by rplevy** [clj-time/clj-time#196]
The [Joda Time project website](http://www.joda.org/joda-time/) declares: "Joda-Time is the de facto standard date and time library for Java. From Java SE 8 onwards, users are asked to migrate to java.time (JSR-310)."

Are there any plans for clj-time to move to JSR-310, or is this a task to be taken up by a different library?

**Comment by michaelklishin** [comment]
This would require JDK 8. Once that version can be required, I see no reason to not use JSR-310. That time may be a good year away, though, and Joda Time generally works pretty well today.

**Comment by rplevy** [comment]
I'm a big fan of Joda Time too, and clj-time makes it even better, which is why I'm asking! That sounds like a good plan.

**Comment by ndpar** [comment]
In case anyone is interested, I made a prototype of clj-time based on Java Time (JDK 8): https://github.com/ndpar/clj-time

**Comment by seancorfield** [comment]
That's an amazing piece of work!

I'll give this a serious look in the next few days. We've started to migrate to Java Time at work and so far we've been using raw Java interop but this would make our lives much easier. 

It will definitely be a breaking change (requiring Java 8 alone is a breaking change, let alone some API changes) but I definitely think this is worth pursuing.

**Comment by michaelklishin** [comment]
Thank you, @ndpar.

**Comment by seancorfield** [comment]
Having reviewed this in a bit more detail, I see some function name changes (`with-time-at-start-of-day` becomes `at-start-of-day`) and removals (`date-midnight`) which would also cause breakage, but I'd like to discuss what it might take to move forward with this...

First of all, what are the pros and cons of dropping Java 6 / Java 7 support?

Second, how much disruption are the API changes likely to be for folks in the real world, and what could / should we do to mitigate them if we move forward with this?

Third, philosophically, how would we want to handle such a big change? Creating a maintenance branch for future 0.x releases based on Joda Time and then work with @ndpar to merge his changes in for a 1.0.0-SNAPSHOT release that would represent the way forward with `clj-time`?

**Comment by seancorfield** [comment]
My answers to these questions:

1. We've been on Java 8 for so long at World Singles that I have no personal interest in Java 7 support. It's been over a year since it got a public update (those ended in April 2015). However, even Oracle's first tier of paid support -- Premier -- covers Java 7 until July 2019 so we should expect corporate users of `clj-time` to want Java 7 supported for a while. I think we can reasonably do that with continued 0.x maintenance releases.

2. I'd like to minimize the renaming / removal of API functions. I think it's fine to remove something that is deprecated. Dropping `milli` is not fine. We can continue to support that by calculating milliseconds from nanoseconds. Adding `nano` _as well_ is fine. We need to be careful about changing the semantics of functions that accept milliseconds today (just changing those to nanoseconds is not fine: that will silently change the meaning of code!). Renaming `with-time-at-start-of-day` to `at-start-of-day` is not fine. When we deprecated `today-at-midnight`, we told people to use `with-time-at-start-of-day` so we shouldn't just remove that (by renaming it) _Note: we forgot to call `deprecated` inside `today-at-midnight` when we did that!_. If we don't like the name, deprecate it and introduce a better name. Similarly, we can't just remove `date-midnight`, even tho' there's no direct Java Time equivalent of the underlying Joda Time DateMidnight class. We can give it some appropriate implementation and deprecate it. We also can't just remove the predicates that test whether values are years, months, days etc -- we need to find a way to implement them, even if we deprecate them.

3. I already made my suggestion :)

**Comment by ndpar** [comment]
All valid points!

Just a small clarification: I dropped/commented out the functions which I personally never used, like `-midnight-`, `years?` etc. From library maintenance perspective we should treat them the way you described, but I would leave the implementation to the people who actually use them. I also agree with minimizing renaming.

The only thing I would argue with is keeping millis in method signatures. Switching to nanos definitely changes the semantics, but sooner or later you have to do it anyway because in a couple of years nobody will think in terms of millis. Since switching from Joda to Java is pretty significant (and essentially not backward compatible) change, it's probably the best time to do it now. It will be more painful later. Or we can try to find the way to use both and deprecate/remove millis from the signatures in the future releases.

**Comment by seancorfield** [comment]
Fair enough on the first paragraph. The reason I raised the issue in the second paragraph is this:

https://github.com/ndpar/clj-time/commit/aa68019e426206d01e5f0441b411f7fcfbfe0d95#diff-99be676f6416ae7390de164e09df814fL324

So `(date-time 2000 6 20 11 10 30 400)` will _silently_ mean something different (because 400 will be treated as nanos instead of millis) so anyone using the 7-arity of the date/time constructors will get an "incorrect" value, compared to what they expect. I'd argue for keeping the semantics the same as today (i.e., 400 would be treated as 400,000 nanos), and possibly adding an 8-arity with millis and then nanos-within-that-milli value, so `(date-time 2000 6 20 11 10 30 0 400)` would be 400 nanos into the second and `(date-time 2000 6 20 11 10 30 1 400)` would be 1,400 nanos into the second. Does that make my concern clearer?

**Comment by ndpar** [comment]
I saw your concern. It makes sense if you are aiming for drop-in replacement. And I'm totally fine with keeping the signature as is.

**Comment by ndpar** [comment]
Sean, I've been thinking about your suggesting to keep millis in the constructor signatures and calculate `milli` from `nano`. Here is the problem. Suppose we keep 7-arity constructor with milli, i.e. in

    (let [d (date-time 2000 6 20 11 10 30 400)] …

400 means 400 millis. Internally we will call Java constructor with 4E8 nanos, because there is no millis in JDK anymore. Now, what should `(milli d)` and `(nano d)` return? 400 and 0? It's not the only choice. 399 and 1E6 is technically correct, too. If you start calculating millis from nanos and vice versa, then `milli` and `nano` are not _independent_ any more.

If you think that 400 and 0 looks better, consider the problem with 8-arity constructor. The following constructors will (should) return the same date-time instant

    (date-time 2000 6 20 11 10 30 400)
    (date-time 2000 6 20 11 10 30 400 0)
    (date-time 2000 6 20 11 10 30 399 1E6)

Do we still prefer 400 and 0 as a result of `milli` and `nano` functions?

This is the reason why Java "silently" replaced Joda's millis with nanos: to prevent this ambiguity. Although I see your concern about changing semantics, I don't see how we can keep it.

**Comment by seancorfield** [comment]
You absolutely cannot silently change the semantics of existing code. That's really horrible for developers. You must either _preserve_ the existing semantics and find a way to offer the alternative semantics with alternative signatures, or you must _break_ the code (and offer a new alternative). I feel very, very strongly about that.

The choice here regarding the 7-arity constructor is clear: it must continue to treat the last argument as millis.

For the 8-arity constructor, you can choose a couple of options:

1. Require that either millis or nanos is zero -- i.e., don't allow both.
1. Require that both millis and nanos are 0 .. 999 so combining them is always valid

Or you can remove the 7-arity constructor -- breaking such calls -- and add an 8-arity constructor where the 7th argument indicates what type the 8th argument should be (`:millis` or `:nanos`) allowing developers a compatibility path to convert any code that has a 7-arity constructor.

**Comment by lkrubner** [comment]
I agree with Sean Corfield, regarding the negative consequences of silently changing the meaning of a function call. I hate it when I upgrade to a new version of a library and suddenly all of my code starts to break in strange ways, though I'm (typically) not getting an obvious error message that tells me what changed and where the error is. 

Could you introduce new function names for the functions that rely on nanoseconds? Something like any of these:

(date-time-n 2000 6 20 11 10 30 400)

(date-time-joda 2000 6 20 11 10 30 400)

(nano-time 2000 6 20 11 10 30 400)

**Comment by kadaj** [comment]
Would like to see JSR-310 migration. Nice library btw.

**Comment by lvh** [comment]
Is there interest in doing this in parallel within clj-time (e.g. joda-time as an optional dependency, perhaps with a dynamic var to pick one or the other), or are we assuming clj-time using `java.time` is just a separate project?

(I would like to use java.time because I also only care about JDK8, and I deploy in constrained environments.)

**Comment by michaelklishin** [comment]
I think it should be two branches with public APIs being as close as reasonable.

**Comment by seancorfield** [comment]
clj-time 0.x will continue to be Joda Time; clj-time 1.x will be Java Time / JSR-310. That way important patches can still be accepted on 0.x for folks who need to remain on Joda Time, but clj-time going forward will switch to Java Time.

So what @michaelklishin said.

We just have a few open Qs to answer before we move forward -- per comments above.

**Comment by danielcompton** [comment]
I'd be in favour of the Java 8 version being a separate project (e.g. new artifactID) that is allowed to make breaking changes. Changing the underlying implementation is already an extremely large change, and anyone doing an upgrade is going to need to carefully vet their code regardless if they ever used any of the Java implementation.

This frees the project developers up to make decisions that make sense given the new API without needing to bring legacy baggage across like the millis/nanos changes.

Having a good upgrade guide describing how to move from one to the other would help reduce the switching costs.

(My 2c)

**Comment by seancorfield** [comment]
@danielcompton makes some good points and I'm very torn about this. I sort of like the idea that `clj-time` could transparently switch from Joda Time to Java Time -- and in our code base at World Singles we only have three references to `org.joda.time` types so that transition would be almost painless. The more Java implementation you rely on, the more painful that transition would be.

Even if you don't rely on the Java implementation explicitly, having a library as heavily used as `clj-time` completely change its internal implementation would require users to think very hard about upgrading and re-testing all of their code.

However, having a different project that offered a (potentially) different API is also going to be a huge change for users -- and you'd be "forcing" a lot more source code changes on them in addition to changing the underlying implementation they've been depending on. If they have to make source code changes _as well as switching the library_ then I think you're looking at a very long, slow adoption curve.

I guess this is why we've sat on this issue for nearly 18 months now -- it's a big decision and there are pros and cons on most approaches.

At work, we already mix `date-clj` and `clj-time` (partly because our database code traffics in `java.util.Date` objects) and in some of our newer code we've already started using Java Time directly. It's not horrible to use "as-is" via Java interop, although some small amount of wrapper code would help (for our use cases so far), but at this point for new code I could imagine eschewing both `date-clj` and `clj-time` and just using Java Time directly.

What other Clojure date/time libraries are there out there? How popular are they? Are there any existing Java Time wrappers that people like? In other words, is migrating `clj-time` to Java Time actually going to add value for the community?

**Comment by danielcompton** [comment]
> However, having a different project that offered a (potentially) different API is also going to be a huge change for users -- and you'd be "forcing" a lot more source code changes on them in addition to changing the underlying implementation they've been depending on.

In my mind, I imagined the goal would still be to keep the spirit of the API the same, avoiding any breakage where at all possible, but allowing for changing things where the java.time API made a different decision to Joda Time.

> In other words, is migrating clj-time to Java Time actually going to add value for the community?

It certainly would for me. I'm doing a lot of interop with Java time directly. clj-time has a good reputation in the community, and a good community around it. Bringing that to a Java 8 API would be very valuable IMO.

I also had another thought about making a new major version for clj-time to use java.time. Because it is so widely adopted in the community, if there are any breaking changes at all, it is going to be challenging to use any libs that depend on the Joda version.

**Comment by seancorfield** [comment]
> I also had another thought about making a new major version for clj-time to use java.time. Because it is so widely adopted in the community, if there are any breaking changes at all, it is going to be challenging to use any libs that depend on the Joda version.

That is the most compelling argument I've heard for not doing this with the same group/artifact because we're limited to one version per project. OK, so that argues strongly for a new group/artifact (and this is exactly the sort of thing Rich was talking about in Spec-ulation when he railed against semantic versioning being used to justify breaking changes!).

I think that means we _must_ have a different project, even if the API is going to be kept almost 100% compatible?

Perhaps instead of creating a new incarnation of `clj-time`, we should rally around an existing Java Time wrapper and support that? Perhaps https://github.com/dm3/clojure.java-time?

**Comment by seancorfield** [comment]
I've been doing some more soul-searching around this issue and, in particular, I think these two blog posts by Stephen Colebourne (project lead on Joda Time, spec lead on JSR-310) are very relevant:

* http://blog.joda.org/2009/11/why-jsr-310-isn-joda-time_4941.html
* http://blog.joda.org/2014/11/converting-from-joda-time-to-javatime.html

The second blog shows the conceptual gaps between the two libraries and this has me fairly convinced that a Java Time version of `clj-time` probably ought to have a (substantially?) different API. @danielcompton's observation above highlights why we really can't just have `clj-time 2.0.0` -- we need to use a different artifact to avoid conflicts within a single project that uses third party libraries that depend on `clj-time`'s current API. And, no matter how much we try to maintain the **Clojure** API, we will still be changing the **Java** API of the library -- we can't escape the fact that `clj-time` has a leaky implementation that causes Joda Time types to popup all over code that uses `clj-time`.

So we have to have a new artifact ID, and we will make breaking API changes even if we keep the Clojure API almost identical (and I think we **must** avoid changing the meaning of names even if we force a `:require` change on the namespace name -- the 7-arg `date-time` constructor issue still really bothers me!).

That means users of `clj-time` who want to switch to this new Java Time version much change every file in their project that currently refers to `clj-time` -- we can't provide a drop-in replacement. That's a lot of work for people.

This is all a long-winded way of saying that I'm coming around to opposing this change now.

**Comment by michaelklishin** [comment]
They can do it at their convenience, though, or never switch. Even after this new library will be out, `clj-time` usage won't drop to 0 in a few months anyway.

**Comment by seancorfield** [comment]
Over the last few months at work, we've continued down the path of using `java.time`, either directly or via `clojure.java-time`, and have slowly been replacing our use of `clj-time` (and `date-clj`). I'm really liking `clojure.java-time` as it feels idiomatic while still staying very close to the underlying `java.time` API.

With that experience under my belt, I'm increasingly opposed to the changes proposed in this issue and would prefer to see `clj-time` continue as the _de facto_ standard Clojure wrapper for Joda Time (and not attempt any switch to Java Time).

For folks who _want_ to use `java.time`, I'd encourage them to check out https://github.com/dm3/clojure.java-time (and maybe that can get enough support to become the _de facto_ Clojure wrapper for Java Time?).

I'd like to close this issue. @michaelklishin Thoughts? Any (strong) objections from anyone else in this thread?

**Comment by danielcompton** [comment]
I'm still quite keen to develop a "clj-time2" which adopts as much of the clj-time API as possible, while providing a smooth migration path for people. There is a ton of code using clj-time, and having a mostly drop-in replacement would help companies and libraries start moving towards `java.time`. 

I'm also a little bit concerned about this:

> The current incarnation of the library is quite slow while calling the 2-3 arity zoned-date-time/offset-time/offset-date-time constructors. If you need predictable latency on the first call, please warm the constructors you are going to use by using them in a 'warm-up phase'

which I think makes it hard to see it becoming a de facto wrapper.

**Comment by seancorfield** [comment]
It's an odd comment, I agree, but if you look at the source you can see why -- and it's very specific to a subset of those "constructors" overall use cases, and it only affects the first call, as far as I can tell. It's more likely to be annoying in the REPL, when the library is first loaded, than in any real world application -- and I haven't noticed any performance issues with the library as a whole in the real world.

For some concrete numbers, the first call of those arities of those constructors can take about three seconds and subsequent calls take under one millisecond. Other arities respond quickly, even on the first call (I saw an occasional 30ms first response, followed by regular sub-1ms times, after repeated cold tests in the REPL).

There's an issue open in that repo to add a cljs version so I think if there's demand, that will happen (and, personally, I don't care about cljs :) ).

**Comment by dm3** [comment]
@danielcompton - thanks for expressing your concern. I'll reword the comment to make it clearer that it only applies to the *first ever* call to the function. The worst I saw in the REPL on my 2.8Ghz i7 was ~400ms.

**Comment by seancorfield** [comment]
This is the only open issue in `clj-time` at the moment so I'd like to move to a resolution. As noted, I'm against the change as originally described and would like to close with no action, given that it any Java Time "version of `clj-time`" is going to be a different artifact with a (slightly?) different API...

**Comment by lvh** [comment]
I dunno if you're looking for a vote, but +1 for clj-jtime that looks mostly like clj-time but uses java.time. If someone wants to write a compatibility shim that occupies the same namespace, they can.

**Comment by seancorfield** [comment]
@lvh I'm fine with someone (not me) writing a new library based on Java Time with an appropriate API (that may or may not overlap substantially with clj-time). So a :+1: for that is a :-1: for this proposed change (which would be to swap the implementation and keep the group/artifact ID and API identical). So I read your vote as agreeing with mine: close this issue.

**Comment by danielcompton** [comment]
Yep, I think I agree here. The change as described of upgrading clj-time to use `java.time` is massively breaking so cannot happen. Like @lvh, I'm pretty keen on a clj-time alike for `java.time`, but that can be handled in a separate issue.

**Comment by seancorfield** [comment]
In the absence of further feedback on this issue, after three months, I'm closing it out with no further action.

In summary:
* `clj-time` will remain a Joda Time wrapper with support for Java 7.
* Folks who want a "`clj-time`-compatible" API over Java Time may build such a library independently (with a different-but-similar group/artifact ID).
* Folks who want Java Time but don't mind migrating to a new library/API can do so independently (we're migrating to `clojure.java-time` but there are other options out there).

Thank you everyone for the discussion and the feedback!
